File: NickLodewijks/src/org/uva/sea/ql/ast/expr/Neg.java
package org.uva.sea.ql.ast.expr;

import org.uva.sea.ql.TypeChecker;
import org.uva.sea.ql.ast.ASTNodeVisitor;
import org.uva.sea.ql.ast.Result;
import org.uva.sea.ql.ast.ValueType;

public class Neg extends AbstractUnaryExpr {

	public Neg(Expr expr) {
		super(expr);
	}

	@Override
	public Integer interpret(Context context) {
		return -Math.abs((Integer) expr.interpret(context));
	}

	@Override
	public ValueType type() {
		return ValueType.INTEGER;
	}

	@Override
	public void _accept(ASTNodeVisitor visitor) {
		visitor.visit(this);
	}

	@Override
	public Result validate() {
		Result result;

		result = TypeChecker.checkType(expr, ValueType.INTEGER);
		if (result.isFalse()) {
			return result;
		}

		return Result.TRUE();
	}
}

File: Edward Poot/src/QL/ast/expr/neg.go
package expr

import (
	"math"
	"ql/ast/visit"
)

type Neg struct {
	Value Expr
}

func (n Neg) getValue() Expr {
	return n.Value
}

func (n Neg) Eval() interface{} {
	return int(math.Abs(float64(n.getValue().Eval().(int))) * -1)
}

func (n Neg) Accept(v visit.Visitor) interface{} {
	return v.Visit(n)
}
